lib/docker.js

```diff
@@ -162,40 +167,56 @@ Docker.prototype.checkAuth = function(opts, callback) {
  * @param {Function} callback Callback
  */
 Docker.prototype.buildImage = function(file, opts, callback) {
+  var pack = tar.pack();
+  var content;
+  var self = this;
+
   if (!callback && typeof opts === 'function') {
     callback = opts;
     opts = null;
   }
 
-  var self = this;
-  var optsf = {
-    path: '/build?',
-    method: 'POST',
-    file: file,
-    options: opts,
-    isStream: true,
-    statusCodes: {
-      200: true,
-      500: 'server error'
-    }
-  };
+  function build(file) {
+    var optsf = {
+      path: '/build?',
+      method: 'POST',
+      file: file,
+      options: opts,
+      isStream: true,
+      statusCodes: {
+        200: true,
+        500: 'server error'
+      }
+    };
 
-  if (opts) {
-    if (opts.registryconfig) {
-      optsf.registryconfig = optsf.options.registryconfig;
-      delete optsf.options.registryconfig;
-    }
+    if (opts) {
+      if (opts.registryconfig) {
+        optsf.registryconfig = optsf.options.registryconfig;
+        delete optsf.options.registryconfig;
+      }
 
-    //undocumented?
-    if (opts.authconfig) {
-      optsf.authconfig = optsf.options.authconfig;
-      delete optsf.options.authconfig;
+      //undocumented?
+      if (opts.authconfig) {
+        optsf.authconfig = optsf.options.authconfig;
+        delete optsf.options.authconfig;
+      }
     }
+
+    self.modem.dial(optsf, function(err, data) {
+      callback(err, data);
+    });
   }
 
-  this.modem.dial(optsf, function(err, data) {
-    callback(err, data);
-  });
+  if (file.context) {
```

Comment:
As you want, I like to handle the most simple and common case first. So to avoid a negation, check if `file` is a string. What happens currently when it's none of these cases ?